### PR TITLE
charger: bq25180: Add more options

### PR DESCRIPTION
**Pull request by gumulka**
Extended the charge controller driver with some more of the supported options.

* Set-options for battery target voltage
* Control for recharge threshold
* Voltage threshold for pre-charge to normal charge

Especially setting the target voltage to something different then 4.2V

### CODE REVIEW

File: drivers/charger/charger_bq25180.c [hunk 1]
```diff
@@ -262,6 +331,12 @@ static int bq25180_init(const struct device *dev)
 		return ret;
 	}
 
+	ret = bq25180_set_charge_voltage(dev, cfg->initial_voltage_microvolt);
+	if (ret < 0) {
+		LOG_ERR("Could not set the target voltage.");
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:344]: I think it's worth logging the return code.

File: drivers/charger/charger_bq25180.c [hunk 2]
```diff
@@ -87,6 +91,32 @@ static uint32_t bq25180_ichg_to_ma(uint8_t ichg)
...
+static int bq25180_mv_to_vbatreg(uint32_t voltage_mv, uint8_t *vbat)
+{
+	if (!IN_RANGE(voltage_mv, BQ25180_VOLTAGE_MIN_MV, BQ25180_VOLTAGE_MAX_MV)) {
+		LOG_WRN("charging voltage out of range: %dmV, "
+			"clamping to the nearest limit",
+			voltage_mv);
+	}
+	voltage_mv = CLAMP(voltage_mv, BQ25180_VOLTAGE_MIN_MV, BQ25180_VOLTAGE_MAX_MV);
+
+	*vbat = (voltage_mv - 3500) / 10;
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:116]: Please `#define` the magic numbers here. With those magic numbers defined, you could honestly drop the comment block for the function.

File: drivers/charger/charger_bq25180.c [hunk 3]
```diff
@@ -87,6 +91,32 @@ static uint32_t bq25180_ichg_to_ma(uint8_t ichg)
...
+
+	return 0;
+}
+
+static uint32_t bq25180_vbatreg_to_mv(uint8_t vbat)
+{
+	vbat &= BQ25180_VBAT_MSK;
+
+	return vbat * 10 + 3500;
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:125]: https://github.com/zephyrproject-rtos/zephyr/pull/77087/commits/d4bf5860f226fbcb93c821e0b9f7e240cd87dfa6#r1717211561

File: drivers/charger/charger_bq25180.c [hunk 4]
```diff
@@ -140,6 +170,41 @@ static int bq25180_get_charge_current(const struct device *dev,
 	return 0;
 }
 
+static int bq25180_set_charge_voltage(const struct device *dev, uint32_t const_charge_voltage_ua)
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:181]: should be `const_charge_voltage_uv`

File: drivers/charger/charger_bq25180.c [hunk 5]
```diff
@@ -140,6 +170,41 @@ static int bq25180_get_charge_current(const struct device *dev,
...
+{
+	const struct bq25180_config *cfg = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = bq25180_mv_to_vbatreg(const_charge_voltage_ua / 1000, &val);
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:187]: https://github.com/zephyrproject-rtos/zephyr/pull/77087/commits/d4bf5860f226fbcb93c821e0b9f7e240cd87dfa6#r1717213116

File: drivers/charger/charger_bq25180.c [hunk 6]
```diff
@@ -140,6 +170,41 @@ static int bq25180_get_charge_current(const struct device *dev,
...
+
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_VBAT_CTRL, BQ25180_VBAT_MSK, val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int bq25180_get_charge_voltage(const struct device *dev, uint32_t *const_charge_voltage_ua)
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:200]: https://github.com/zephyrproject-rtos/zephyr/pull/77087/commits/d4bf5860f226fbcb93c821e0b9f7e240cd87dfa6#r1717213116

File: dts/bindings/charger/ti,bq25180.yaml [hunk 7]
```diff
@@ -31,3 +31,10 @@ properties:
 
   constant-charge-voltage-max-microvolt:
     default: 4200000
+
+  precharge-voltage-threshold-microvolt:
+    type: int
```
Comment (rriveramcrus) [inline dts/bindings/charger/ti,bq25180.yaml:36]: If this value can only be 2.8V or 3.0V, then maybe it should be `type: enum` instead.

Comment (gumulka) [reply dts/bindings/charger/ti,bq25180.yaml:36]: I thought about it to, but decided against it, as all other values in https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/battery/battery.yaml are int in uv and wanted to have a somewhat consistent interface. The re-charge voltage is also an enum with either 100 or 200uv below max voltage, but this is already defined in battery.yaml. Maybe one could think about extending the battery.yaml instead of this charger, but I was unsure about that.

File: drivers/charger/charger_bq25180.c (hunk 4 above)
Comment (gumulka) [reply drivers/charger/charger_bq25180.c]: yeah, sorry about that. Forgot it when copy pasting... Fixed now.

File: drivers/charger/charger_bq25180.c (hunk 1 above)
Comment (gumulka) [reply drivers/charger/charger_bq25180.c]: Done

File: drivers/charger/charger_bq25180.c (hunk 2 above)
Comment (gumulka) [reply drivers/charger/charger_bq25180.c]: Done

File: drivers/charger/charger_bq25180.c [hunk 8]
```diff
@@ -87,6 +93,27 @@ static uint32_t bq25180_ichg_to_ma(uint8_t ichg)
...
+	*vbat = (voltage_mv - BQ25180_VOLTAGE_MIN_MV) / BQ25180_FACTOR_VBAT_TO_MV;
+
+	return 0;
+}
+
+static uint32_t bq25180_vbatreg_to_mv(uint8_t vbat)
+{
+	vbat &= BQ25180_VBAT_MSK;
+
+	return vbat * BQ25180_FACTOR_VBAT_TO_MV + BQ25180_VOLTAGE_MIN_MV;
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:123]:
``` c
return (vbat * BQ25180_FACTOR_VBAT_TO_MV) + BQ25180_VOLTAGE_MIN_MV;
```
Some parenthesis would help with readability

Comment (gumulka) [reply drivers/charger/charger_bq25180.c]: No problem :)

File: drivers/charger/charger_bq25180.c [hunk 9]
```diff
@@ -120,17 +125,15 @@ static int bq25183_charge_enable(const struct device *dev, const bool enable)
 	uint8_t value = enable ? 0 : BQ25180_ICHG_CHG_DIS;
 	int ret;
 
-	ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL,
-				     BQ25180_ICHG_CHG_DIS, value);
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL, BQ25180_ICHG_CHG_DIS, value);
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:132]: While an improvement, this change is out of scope of this commit.

File: drivers/charger/charger_bq25180.c [hunk 10]
```diff
@@ -120,17 +125,15 @@ static int bq25183_charge_enable(const struct device *dev, const bool enable)
...
 	if (ret < 0) {
 		return ret;
 	}
 
 	return 0;
 }
 
-static int bq25180_set_charge_current(const struct device *dev,
-				      uint32_t const_charge_current_ua)
+static int bq25180_set_charge_current(const struct device *dev, uint32_t const_charge_current_ua)
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:140]: While an improvement, this change is out of scope of this commit.

File: drivers/charger/charger_bq25180.c [hunk 11]
```diff
@@ -141,17 +144,15 @@ static int bq25180_set_charge_current(const struct device *dev,
 		return ret;
 	}
 
-	ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL,
-				     BQ25180_ICHG_MSK, val);
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL, BQ25180_ICHG_MSK, val);
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:151]: While an improvement, this change is out of scope of this commit.

File: drivers/charger/charger_bq25180.c [hunk 12]
```diff
@@ -141,17 +144,15 @@ static int bq25180_set_charge_current(const struct device *dev,
...
 	if (ret < 0) {
 		return ret;
 	}
 
 	return 0;
 }
 
-static int bq25180_get_charge_current(const struct device *dev,
-				      uint32_t *const_charge_current_ua)
+static int bq25180_get_charge_current(const struct device *dev, uint32_t *const_charge_current_ua)
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:159]: While an improvement, this change is out of scope of this commit.

File: drivers/charger/charger_bq25180.c [hunk 13]
```diff
@@ -202,8 +203,7 @@ static int bq25180_get_charge_voltage(const struct device *dev, uint32_t *const_
 	return 0;
 }
 
-static int bq25180_get_online(const struct device *dev,
-			      enum charger_online *online)
+static int bq25180_get_online(const struct device *dev, enum charger_online *online)
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:210]: While an improvement, this change is out of scope of this commit.

File: drivers/charger/charger_bq25180.c [hunk 14]
```diff
@@ -223,8 +223,7 @@ static int bq25180_get_online(const struct device *dev,
 	return 0;
 }
 
-static int bq25180_get_status(const struct device *dev,
-			      enum charger_status *status)
+static int bq25180_get_status(const struct device *dev, enum charger_status *status)
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:230]: While an improvement, this change is out of scope of this commit.

File: drivers/charger/charger_bq25180.c [hunk 15]
```diff
@@ -321,8 +320,7 @@ static int bq25180_init(const struct device *dev)
 	}
 
 	/* Disable the watchdog */
-	ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_IC_CTRL,
-				     BQ25180_WATCHDOG_SEL_1_MSK,
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, BQ25180_IC_CTRL, BQ25180_WATCHDOG_SEL_1_MSK,
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:327]: While an improvement, this change is out of scope of this commit.

File: dts/bindings/charger/ti,bq25180.yaml [hunk 16]
```diff
@@ -28,3 +28,6 @@ properties:
       The value specified will be rounded down to the closest implemented
       value. If set to 0 (default) skip setting the charge current value at
       driver initialization.
+
+  constant-charge-voltage-max-microvolt:
+    default: 4200000
```
Comment (rriveramcrus) [inline dts/bindings/charger/ti,bq25180.yaml:33]: It might be safer to make this a required rather than provide a default. 4.2V is safe for most lithium batteries, but not all. Given the nature of lithium batteries, safer is usually better :)

File: dts/bindings/charger/ti,bq25180.yaml (hunk 7 above)
Comment (rriveramcrus) [reply dts/bindings/charger/ti,bq25180.yaml:36]: Thanks for clearing this up. I am in favor of how you have implemented this change.

File: dts/bindings/charger/ti,bq25180.yaml (hunk 16 above)
Comment (gumulka) [reply dts/bindings/charger/ti,bq25180.yaml:33]: Can do, I just used what was already the default value for the device. (And the reason for this whole pull request, because my battery does need a lower voltage.)

File: drivers/charger/charger_bq25180.c [hunk 17]
```diff
@@ -68,7 +72,8 @@ static int bq25180_ma_to_ichg(uint32_t current_ma, uint8_t *ichg)
 {
 	if (!IN_RANGE(current_ma, BQ25180_CURRENT_MIN_MA, BQ25180_CURRENT_MAX_MA)) {
 		LOG_WRN("charging current out of range: %dmA, "
-			"clamping to the nearest limit", current_ma);
+			"clamping to the nearest limit",
+			current_ma);
```
[the remark this answers is not in the document]
Comment (gumulka) [reply drivers/charger/charger_bq25180.c]: Removed all these changes.

File: drivers/charger/charger_bq25180.c [hunk 18]
```diff
@@ -17,31 +17,34 @@
...
+#define BQ25180_IC_CTRL_VRCH_100                0x00
+#define BQ25180_IC_CTRL_VRCH_200                BIT(5)
+#define BQ25180_IC_CTRL_VRCH_MSK                BIT(5)
+#define BQ25180_WATCHDOG_SEL_1_MSK              GENMASK(1, 0)
+#define BQ25180_WATCHDOG_DISABLE                0x03
+#define BQ25180_DEVICE_ID_MSK                   GENMASK(3, 0)
+#define BQ25180_DEVICE_ID                       0x00
+#define BQ25180_SHIP_RST_EN_RST_SHIP_MSK        GENMASK(6, 5)
+#define BQ25180_SHIP_RST_EN_RST_SHIP_ADAPTER    0x20
+#define BQ25180_SHIP_RST_EN_RST_SHIP_BUTTON     0x40
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c]:
``` c
#define BQ25180_IC_CTRL_VRCH_100                0x00
#define BQ25180_WATCHDOG_DISABLE 0x03	#define BQ25180_IC_CTRL_VRCH_200                BIT(5)
#define BQ25180_DEVICE_ID_MSK GENMASK(3, 0)	#define BQ25180_IC_CTRL_VRCH_MSK                BIT(5)
```
Only these lines should be introduced. Using `git clang-format` would only format your staged changes and avoid spurious edits.

Comment (gumulka) [reply drivers/charger/charger_bq25180.c]: Changed :)

File: drivers/charger/charger_bq25180.c [hunk 19]
```diff
@@ -87,6 +93,27 @@ static uint32_t bq25180_ichg_to_ma(uint8_t ichg)
...
 }
 
+static int bq25180_mv_to_vbatreg(uint32_t voltage_mv, uint8_t *vbat)
+{
+	if (!IN_RANGE(voltage_mv, BQ25180_VOLTAGE_MIN_MV, BQ25180_VOLTAGE_MAX_MV)) {
+		LOG_WRN("charging voltage out of range: %dmV, "
+			"clamping to the nearest limit",
+			voltage_mv);
+	}
+	voltage_mv = CLAMP(voltage_mv, BQ25180_VOLTAGE_MIN_MV, BQ25180_VOLTAGE_MAX_MV);
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:111]: Minor nit, it would be nicer to clamp the maximum to what is provided by `constant-charge-voltage-max-microvolt`. This would ensure the regulation voltage target never exceeds what is set higher than the maximum set in DT.

File: drivers/charger/charger_bq25180.c [hunk 20]
```diff
@@ -334,6 +338,20 @@ static int bq25180_init(const struct device *dev)
 		return ret;
 	}
 
+	if (cfg->recharge_voltage_microvolt > 0) {
+		if (cfg->initial_voltage_microvolt - cfg->recharge_voltage_microvolt > 100000) {
```
Comment (rriveramcrus) [inline drivers/charger/charger_bq25180.c:346]: I would appreciate parenthesis here for readability

**Comment by rriveramcrus** [comment]
After these last two comments are address, I'm ready to approve. @GRobertZieba, would you be able to take a look as the 2nd reviewer?

File: drivers/charger/charger_bq25180.c (hunk 19 above)
Comment (gumulka) [reply drivers/charger/charger_bq25180.c]: Also renamed the variable from `initial_voltage_microvolt` to `max_voltage_microvolt`

**Review by rriveramcrus — APPROVED** [review]

**Review by RobertZ2011 — APPROVED** [review]

**Comment by gumulka** [comment]
@rriveramcrus @GRobertZieba is there something that I can do for it to be merged?

**Comment by rriveramcrus** [comment]
> @rriveramcrus @GRobertZieba is there something that I can do for it to be merged?

Oh I see what's going on @RobertZ2011 has a new account that does not have collaborator status, so we will need to take care of that. In the meantime let me see if I can find someone else to take a look. No action is need from you, @gumulka.

**Review by teburd — APPROVED** [review]
